File: python/src/gibberish_rewriter/win/input_builder.py
"""Builds the event arrays InputSender passes to SendInput. Pure, so it is unit-tested.

Mirrors ``csharp/src/GibberishRewriter.App/Win/InputBuilder.cs``.
"""

from __future__ import annotations

from collections.abc import Iterator

from gibberish_rewriter.core.platform import Shortcut

from gibberish_rewriter.win import win32
from gibberish_rewriter.win.win32 import INPUT, INPUT_KEYBOARD, KEYBDINPUT, InputUnion

EXTRA_INFO = 0x47525752
"""dwExtraInfo on every event this app sends ("GRWR"); KeyboardHook recognizes the app's own input by it."""

CHUNK_SIZE = 50
"""SendInput is called with at most this many events at a time."""

VK_BACK = 0x08
VK_SHIFT = 0x10
VK_CONTROL = 0x11
VK_SPACE = 0x20
VK_INSERT = 0x2D
VK_C = 0x43
VK_V = 0x56
VK_Z = 0x5A
VK_LWIN = 0x5B


def key(vk: int, up: bool) -> INPUT:
    """One virtual-key event."""
    flags = (win32.KEYEVENTF_KEYUP if up else 0) | (
        win32.KEYEVENTF_EXTENDEDKEY if _is_extended(vk) else 0
    )
    return INPUT(
        type=INPUT_KEYBOARD,
        u=InputUnion(ki=KEYBDINPUT(wVk=vk, dwFlags=flags, dwExtraInfo=EXTRA_INFO)),
    )


def unicode(unit: int, up: bool) -> INPUT:
    """One KEYEVENTF_UNICODE event. C# takes a ``char``; ``unit`` is that UTF-16 code unit as a number, because
    Python's ``str`` iterates code points rather than code units."""
    flags = win32.KEYEVENTF_UNICODE | (win32.KEYEVENTF_KEYUP if up else 0)
    return INPUT(
        type=INPUT_KEYBOARD,
        u=InputUnion(ki=KEYBDINPUT(wScan=unit, dwFlags=flags, dwExtraInfo=EXTRA_INFO)),
    )


def key_press(vk: int) -> list[INPUT]:
    """One press (down, up) of a virtual key: the CapsLock replay or the 0xE8 mask key."""
    return [key(vk, False), key(vk, True)]


def backspaces(count: int) -> list[INPUT]:
    events: list[INPUT] = []
    for _ in range(count):
        events.append(key(VK_BACK, False))
        events.append(key(VK_BACK, True))
    return events


def text(value: str) -> list[INPUT]:
    """Each UTF-16 unit as a KEYEVENTF_UNICODE down and up.

    A character outside the basic plane is two units, exactly as it is two ``char`` values in C#.
    """
    units = value.encode("utf-16-le", "surrogatepass")
    events: list[INPUT] = []
    for index in range(0, len(units), 2):
        unit = units[index] | (units[index + 1] << 8)
        events.append(unicode(unit, False))
        events.append(unicode(unit, True))
    return events


def shortcut(which: Shortcut, console: bool) -> list[INPUT]:
    """Ctrl+C / Ctrl+V / Ctrl+Z, or Ctrl+Insert / Shift+Insert in console windows."""
    if which is Shortcut.COPY:
        modifier, trigger = (VK_CONTROL, VK_INSERT) if console else (VK_CONTROL, VK_C)
    elif which is Shortcut.PASTE:
        modifier, trigger = (VK_SHIFT, VK_INSERT) if console else (VK_CONTROL, VK_V)
    else:
        modifier, trigger = VK_CONTROL, VK_Z
    return [key(modifier, False), key(trigger, False), key(trigger, True), key(modifier, True)]


def win_space() -> list[INPUT]:
    """Win+Space, the layout-switch fallback."""
    return [key(VK_LWIN, False), key(VK_SPACE, False), key(VK_SPACE, True), key(VK_LWIN, True)]


def chunks(events: list[INPUT]) -> Iterator[list[INPUT]]:
    for start in range(0, len(events), CHUNK_SIZE):
        yield events[start : min(start + CHUNK_SIZE, len(events))]


def _is_extended(vk: int) -> bool:
    return vk in (VK_INSERT, VK_LWIN)

File: csharp/src/GibberishRewriter.App/Win/InputBuilder.cs
using static GibberishRewriter.App.Win.NativeMethods;
using Shortcut = GibberishRewriter.Core.Shortcut;

namespace GibberishRewriter.App.Win;

/// <summary>Builds the event arrays InputSender passes to SendInput. Pure, so it is unit-tested.</summary>
internal static class InputBuilder
{
    /// <summary>dwExtraInfo on every event this app sends ("GRWR"); KeyboardHook recognizes the app's own input by it.</summary>
    public const uint ExtraInfo = 0x47525752;

    /// <summary>SendInput is called with at most this many events at a time.</summary>
    public const int ChunkSize = 50;

    public const ushort VkBack = 0x08;
    public const ushort VkShift = 0x10;
    public const ushort VkControl = 0x11;
    public const ushort VkSpace = 0x20;
    public const ushort VkInsert = 0x2D;
    public const ushort VkC = 0x43;
    public const ushort VkV = 0x56;
    public const ushort VkZ = 0x5A;
    public const ushort VkLWin = 0x5B;

    public static INPUT Key(ushort vk, bool up) => new()
    {
        type = INPUT_KEYBOARD,
        u = new InputUnion
        {
            ki = new KEYBDINPUT
            {
                wVk = vk,
                dwFlags = (up ? KEYEVENTF_KEYUP : 0) | (IsExtended(vk) ? KEYEVENTF_EXTENDEDKEY : 0),
                dwExtraInfo = (UIntPtr)ExtraInfo,
            },
        },
    };

    public static INPUT Unicode(char c, bool up) => new()
    {
        type = INPUT_KEYBOARD,
        u = new InputUnion
        {
            ki = new KEYBDINPUT
            {
                wScan = c,
                dwFlags = KEYEVENTF_UNICODE | (up ? KEYEVENTF_KEYUP : 0),
                dwExtraInfo = (UIntPtr)ExtraInfo,
            },
        },
    };

    /// <summary>One press (down, up) of a virtual key: the CapsLock replay or the 0xE8 mask key.</summary>
    public static INPUT[] KeyPress(ushort vk) => [Key(vk, false), Key(vk, true)];

    public static INPUT[] Backspaces(int count)
    {
        var events = new INPUT[count * 2];
        for (var i = 0; i < count; i++)
        {
            events[i * 2] = Key(VkBack, false);
            events[i * 2 + 1] = Key(VkBack, true);
        }
        return events;
    }

    /// <summary>Each UTF-16 unit as a KEYEVENTF_UNICODE down and up.</summary>
    public static INPUT[] Text(string text)
    {
        var events = new INPUT[text.Length * 2];
        for (var i = 0; i < text.Length; i++)
        {
            events[i * 2] = Unicode(text[i], false);
            events[i * 2 + 1] = Unicode(text[i], true);
        }
        return events;
    }

    /// <summary>Ctrl+C / Ctrl+V / Ctrl+Z, or Ctrl+Insert / Shift+Insert in console windows.</summary>
    public static INPUT[] Shortcut(Shortcut shortcut, bool console)
    {
        var (modifier, key) = shortcut switch
        {
            Core.Shortcut.Copy => console ? (VkControl, VkInsert) : (VkControl, VkC),
            Core.Shortcut.Paste => console ? (VkShift, VkInsert) : (VkControl, VkV),
            _ => (VkControl, VkZ),
        };
        return [Key(modifier, false), Key(key, false), Key(key, true), Key(modifier, true)];
    }

    /// <summary>Win+Space, the layout-switch fallback.</summary>
    public static INPUT[] WinSpace() => [Key(VkLWin, false), Key(VkSpace, false), Key(VkSpace, true), Key(VkLWin, true)];

    public static IEnumerable<INPUT[]> Chunks(INPUT[] events)
    {
        for (var start = 0; start < events.Length;)
        {
            var end = Math.Min(start + ChunkSize, events.Length);
            // A surrogate pair is four events and ChunkSize is not a multiple of four, so a chunk boundary can fall
            // inside one. Windows wants a pair in one call, so step back to where the pair starts.
            while (end > start + 1 && end < events.Length && IsHighSurrogate(events[end - 1]))
            {
                end--;
            }
            yield return events[start..end];
            start = end;
        }
    }

    private static bool IsHighSurrogate(INPUT input) =>
        (input.u.ki.dwFlags & KEYEVENTF_UNICODE) != 0 && char.IsHighSurrogate((char)input.u.ki.wScan);

    private static bool IsExtended(ushort vk) => vk is VkInsert or VkLWin;
}
